File: problem-solving/lc-70-easy-climbing-stairs/3-bottom-up-optimized.py
"""
https://leetcode.com/problems/climbing-stairs/

Complexity
----------
Time: O(N)
Space: O(1)
"""

class Solution:
    def climbStairs(self, n: int) -> int:
        away = 1
        away2 = 2
        res = n
        for v in range(3, n + 1):
            res = away + away2
            away = away2
            away2 = res
        return res

File: problem-solving/lc-70-easy-climbing-stairs/3-bottom-up-optimized.go
/**
 * https://leetcode.com/problems/climbing-stairs/
 * 
 * Complexity
 * ----------
 * Time: O(N)
 * Space: O(1)
 */

func climbStairs(n int) int {
    if n < 4 {
        return n
    }
    var count int
    oneAway := 3
    twoAway := 2
    for i := 4; i <= n; i++ {
        count = oneAway + twoAway
        twoAway = oneAway
        oneAway = count
    }
    return count
}
